src/Avalonia.Controls/TransitioningContentControl.cs

```diff
@@ -162,8 +182,10 @@ private void HideOldPresenter()
         var oldPresenter = _isFirstFull ? _presenter2 : Presenter;
         if (oldPresenter is not null)
         {
-            oldPresenter.Content = null;
-            oldPresenter.IsVisible = false;
+          var oldcontent = oldPresenter.Content;
+          oldPresenter.Content = null;
+          oldPresenter.IsVisible = false;
+          OldContent = oldcontent;
```

Comment:
Couldn't you just handle control.PropertyChanged event, check for Content property and get OldValue from the arguments?
Otherwise, OldContent property feels weird + will hold object in memory more than it should.

Reply:
I was just following all the places in the original design where the `Content` of the "old" presenter was being set to null, which happens in 2 places, not 1. So from my understanding merely handling `PropertyChanged` would not be enough as that would account for just `UpdateContent` path ignoring the `ArrangeOverride` path that also calls `HideOldPresenter`.

> will hold object in memory more than it should

I don't like it too. But how can this be avoided with direct properties?

Reply:
As far as holding object in memory, getter could be modified as such, with documentation explaining that the value is available momentarily:

```cs
    get {
      var v = _OldContent;
      _OldContent = null;
      return v;
    }
```

Reply:
And regarding

> Couldn't you just handle control.PropertyChanged event, check for Content property and get OldValue from the arguments?

If you mean something like this:

```cs
    if (change.Property == ContentProperty) {
      UpdateContent(true);
      //report change.OldValue somehow
      return;
    }
```

Then it does not work because of how `TransitioningContentControl` operates. I tried that approach and started getting weird exceptions related to the fact that I disposed of OldContent in my ViewModel prematurely. So, it looks like the OldContent needs to be set really in the 2 places where originally OldPresenter's content is being set to null - and only there.

Reply:
Have you tried handling OnDetachedFromVisualTree method overload in your content? That's a typical solution for control lifetime - initialize in attached-to-visualtree and dispose in detached-to-visualtree.

Reply:
The problem is that often the "Content" is not an Avalonia object at all, let alone a control. It is an arbitrary DotNet class where there is a `TransitioningContentControl.ContentTemplate` that describes how to actually present such a Content. Hence the need to expose OldContent in a data-bindable-friendly fashion on the `TransitioningContentControl`

Reply:
@maxkatz6 

1. I updated the PR to also account for the missed case when the Old Content goes "out of scope" so to speak and is no longer needed when it happens _during the transition_. It is important to account of all such cases so that Old Content changes can be properly observed by the VM and dealt with as needed.
2. Regarding the backing field for the OldContent property: I do not like it too, but how else can this be addressed? Perhaps the best way would be to instead of doing:

```cs
var oldcontent = oldPresenter.Content;
oldPresenter.Content = null;
OldContent = oldcontent;
```

do this:

```cs
var oldcontent = oldPresenter.Content;
oldPresenter.Content = null;
OldContent = oldcontent;
OldContent = null; //We have signaled to the VM that the old content was no longer needed, so we can now set it to null alltogether - it should have already been taken care of by whoever handled the above line
```

Reply:
Not sure why seemingly unrelated tests failed. I am new to this, so if anyone can clarify, that would be nice.